airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py

```diff
@@ -642,38 +642,37 @@ def adopt_launched_task(
         del tis_to_flush_by_key[ti_key]
         self.running.add(ti_key)
 
-    def _adopt_completed_pods(self, kube_client: client.CoreV1Api) -> None:
+    @provide_session
+    def _delete_orphaned_completed_pods(self, session: Session = NEW_SESSION) -> None:
         """
-        Patch completed pods so that the KubernetesJobWatcher can delete them.
+        Delete orphaned completed pods with completed TaskInstances.
 
-        :param kube_client: kubernetes client for speaking to kube API
+        Pods that have reached the Completed status are usually deleted by the scheduler to which
+        they are attached. In case when the scheduler crashes, there is no one to delete these
+        pods. Therefore, they are deleted from another scheduler using this function.
         """
+        from airflow.jobs.job import Job, JobState
+
         if TYPE_CHECKING:
-            assert self.scheduler_job_id
+            assert self.kube_scheduler
 
-        new_worker_id_label = self._make_safe_label_value(self.scheduler_job_id)
-        query_kwargs = {
-            "field_selector": "status.phase=Succeeded",
-            "label_selector": (
-                "kubernetes_executor=True,"
-                f"airflow-worker!={new_worker_id_label},{POD_EXECUTOR_DONE_KEY}!=True"
-            ),
-        }
+        alive_schedulers_ids = session.scalars(
+            select(Job.id).where(Job.job_type == "SchedulerJob", Job.state == JobState.RUNNING)
+        ).all()
+        labels = ["kubernetes_executor=True", f"{POD_EXECUTOR_DONE_KEY}!=True"]
+        for alive_scheduler_id in alive_schedulers_ids:
+            labels.append(f"airflow-worker!={self._make_safe_label_value(str(alive_scheduler_id))}")
+
+        query_kwargs = {"field_selector": "status.phase=Succeeded", "label_selector": ",".join(labels)}
         pod_list = self._list_pods(query_kwargs)
         for pod in pod_list:
-            self.log.info("Attempting to adopt pod %s", pod.metadata.name)
             from kubernetes.client.rest import ApiException
 
             try:
-                kube_client.patch_namespaced_pod(
-                    name=pod.metadata.name,
-                    namespace=pod.metadata.namespace,
-                    body={"metadata": {"labels": {"airflow-worker": new_worker_id_label}}},
-                )
+                self.kube_scheduler.delete_pod(pod_name=pod.metadata.name, namespace=pod.metadata.namespace)
+                self.log.info("Orphaned completed pod %s has been deleted", pod.metadata.name)
             except ApiException as e:
-                self.log.info("Failed to adopt pod %s. Reason: %s", pod.metadata.name, e)
-            ti_id = annotations_to_key(pod.metadata.annotations)
-            self.running.add(ti_id)
```

Comment:
We need to gracefully free the executor running slot from the concerned scheduler. Otherwise it leads to leak.

Reply:
Can you clarify what you mean by a 'concerned scheduler'? I understand that this function is needed when one of the schedulers is down (i.e., killed by OOM). When your pod is completed, and if the scheduler that it was attached to is down, the TaskInstance corresponding to this pod is not in any running set of the rest of the schedulers. And you don't have to adopt it (or add it to the running set of any alive scheduler) because it's already been completed. Perhaps this comment will provide more information for you: link (https://github.com/apache/airflow/pull/35800#issuecomment-1845545959).

Are you referring to the situation when the scheduler missed its heartbeat and was announced as 'failed' by another scheduler? Shouldn't we fail loudly in this situation? Should we really mask the problem with skipping heartbeats this way?

Reply:
Yes. I am referring to the missing heartbeat scenario. This will cause the executor running slot leak due to its been deleted by other scheduler due to its current scheduler heartbeat is delayed.

Reply:
Ok we'll try to refactor this function to it's initial state + change kube selector to adopt pods only from failed scheduler + add continue to the except block. We will run this patch on our airflow setup and will comeback with the results.

Reply:
The original adoption code will adopt the pods from other schedulers including the failed ones. Once they are adopted. the watcher will get a successful event. and then the executor gracefully deletes the pod including freeing the running queue slot.  I would request you to try the latest main code to see if that fixes this problem. The latest main code has executor running slot leak fix.

Reply:
Adoption should be performed only on pods that bound to failed schedulers. It should not be performed on pods that are bound to normally working schedulers.

Reply:
> Adoption should be performed only on pods that are bound to failed schedulers. It should not be performed on pods that are bound to normally working schedulers.

Why would that apply to completed PODs as well? What problem would it cause? 

I understand that what we are trying to do here is not to `adopt` those pods, but to `delete`  completed ones (this is the gist of that change as I see it). There is a bit of difference in `adopting` and taking care about the POD, than `deleting` the POD (which is the last thing that happens to it). As I see see it, yes, adoption means "I will take care about this POD in the future", where "deletion" means "let's delete this pod and not care about it any more". Those are two completely differnt cases.

Or am I missing something?

Reply:
I think I misinterpreted "missed heartbeat" situation. I need a little more time to properly think about it.

Reply:
Not trying to muddy the waters here @dirrao but I am curious about your comment that scheduler 1 gets DELETED event when pod adopted.... if the pod were actually deleted, wouldn't there also be a DELETED event? (though for a different reason)

Also, there's something messed up in the main code that you quoted earlier in this thread.  you have logic like this
```
if event == deleted:
    ...
elif ...:
    if event == deleted
```

that second if event == deleted will never be reached.

Reply:
Right now, the Airflow K8s watcher is subscribed to pod events with a scheduler id label selector. When the scheduler ID is changed in pod labels during adoption of queued tasks or completed tasks, the original scheduler gets a DELETE event but it's not deleted. The way to identify the deletion of a pod is based on **pod.metadata.deletion_timestamp**. 

```
        if event["type"] == "DELETED" and not pod.metadata.deletion_timestamp:
               print("pod is adopted")
        elif event["type"] == "DELETED" and pod.metadata.deletion_timestamp:
               print("pod is deleted")
```